turbinia/workers/analysis/postgresql_acct.py

```diff
@@ -0,0 +1,235 @@
+        hashnames, timeout=timeout)
+    output_evidence.text_data = report
+    result.report_data = report
+    result.report_priority = priority
+
+    # Write the report to the output file.
+    with open(output_file_path, 'wb') as fh:
+      fh.write(output_evidence.text_data.encode('utf8'))
+      fh.write('\n'.encode('utf8'))
+
+    # Add the resulting evidence to the result object.
+    result.add_evidence(output_evidence, evidence.config)
+    result.close(self, success=True, status=summary)
+
+    return result
+
+  def _collect_conf_files(self, evidence):
+    """Extract artifacts using image_export.
+
+    Args:
+        evidence (Evidence object):  The evidence to process
+    Returns:
+        location (str): The file path to the extracted evidence.
+        number of artifacts (int): The number of files extracted.
+    """
+    try:
+      collected_artifacts = extract_files(
+          file_name=_PG_CONF_NAME,
+          disk_path=evidence.local_path, output_dir=os.path.join(
+              self.output_dir, 'artifacts'), credentials=evidence.credentials)
+    except TurbiniaException as e:
+      raise TurbiniaException(
+          'artifact extraction failed: {0:s}'.format(str(e)))
+
+    # Extract base dir from our list of collected artifacts
+    location = os.path.dirname(collected_artifacts[0])
+
+    return (location, len(collected_artifacts))
+
+  def _extract_data_dir(self, location, result):
+    """Attempts to extract the data_directory value
+
+    Args:
+      location : Directory of files extracted from the disk.
+      result (TurbiniaTaskResult): Used to log messages.
+
+    Returns:
+      data_dirs (list): List of locations of pgsql databases
+    """
+    data_dirs = []
+    for dirs, _, _ in os.walk(location):
+      try:
+        grep = subprocess.run(
+            ['grep', r'data_directory',
+             os.path.join(dirs, _PG_CONF_NAME)], check=False, text=True,
+            capture_output=True)
+        if grep.returncode != 0:
+          continue
+
+        for directive in grep.stdout.strip().split('\n'):
+          parts = directive.split("'")
+          if len(parts) == 3:
+            data_dirs.append(parts[1])
+          else:
+            result.log(
+                'Unable to parse data_dir directive: {0:s}'.format(directive))
+      except subprocess.CalledProcessError as e:
+        raise TurbiniaException(
+            'Unable to grep Postgres config file: {0:s}'.format(str(e)))
+
+    result.log('Returning {}'.format(data_dirs))
+    return data_dirs
+
+  def _extract_creds(self, locations, evidence):
+    """Attempts to extract raw encrypted credentials from the database
+
+    Args:
+      locations : Dead disk database dirs.
+      evidence (Evidence object):  The evidence to process
+
+    Returns:
+      hashnames (dict): Dict mapping hash back to username.
+    """
+    hashnames = {}
+    for location in locations:
+      dir = os.path.normpath(evidence.local_path + location)
```

Comment:
I actually spent an entire day debugging why this wasn't working because I was using os.path.join().
Here's a nice little sentence buried deep in the documentation:
"If a component is an absolute path, all previous components are thrown away and joining continues from the absolute path component."
Thus `os.path.join('/path/to/evidence', '/var/log/whatever')` will yield `/var/log/whatever` and silently throw away all previous arguments.

Reply:
Woah, that's bizarre.  That seems like a bug to me -- I wonder why it's like that.